### PR TITLE
Winrm kerberos authentication support

### PR DESCRIPTION
**Pull request by smashery**
Relies on #16700

This PR implements Kerberos login for WinRM. Currently it's used on the `winrm_cmd` module for individual commands, but I intend to support sessions prior to 

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use winrm_cmd`
- [ ] Set `auth` to `kerberos`, and then set `rhost`, `username`, `password`, `domain`, `domaincontrollerip` and `cmd`
- [ ] Verify that the authentication process is navigated, and commands work against:
- [ ] Windows 2008 DC (`rport` should be `80`). Note this does not support AES, so everything should be RC4
- [ ] Windows 2008 Client (`rport` should be `80`)
- [ ] Windows 2022 (Should be AES256 by default)
- [ ] Any other versions in between these that you want to try

## Prior to removing draft status

- Wait for #16700 to land (this should fix most of the unit test issues, but I expect some additional ones will need looking at)
- Decide on how to launch WinRM sessions through this mechanism (should it be the `winrm_login` module, or a new one)?

### CODE REVIEW

**Comment by adfoster-r7** [comment]
Verified as working against DC with windows 2016

```
msf6 auxiliary(scanner/winrm/winrm_cmd) > run 192.168.123.13 domain=adf3.local username=Administrator password=p4$$w0rd Winrmauth=kerberos domaincontrollerrhost=192.168.123.13 winrmrhostname=dc3.adf3.local
[*] 192.168.123.13:88 - Received a valid TGT-Response
[+] 192.168.123.13:88 - Received a valid TGS-Response
[*] 192.168.123.13:88 - TGS MIT Credential Cache saved to /Users/adfoster/.msf4/loot/20220707153004_default_192.168.123.13_mit.kerberos.cca_699523.bin
[+] 192.168.123.13:88 - Received AP-REQ. Extracting session key...

Windows IP Configuration

   Host Name . . . . . . . . . . . . : dc3
   Primary Dns Suffix  . . . . . . . : adf3.local
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : adf3.local

```

File: lib/rex/proto/kerberos/model.rb [hunk 1]
```diff
@@ -8,14 +8,17 @@ module Model
...
 
-        AS_REQ = 10
-        AS_REP = 11
-        TGS_REQ = 12
-        TGS_REP = 13
-        KRB_ERROR = 30
-        TICKET = 1
-        AUTHENTICATOR = 2
-        AP_REQ = 14
+        #module MessageType
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/model.rb:11]: @smashery Sounds good to me to namespace this; Are we good to put this in a separate PR, or were there technical reasons for not wanting to add the namespace 👀

File: lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            flags |= GSS_ACCEPTOR_SUBKEY if use_acceptor_subkey
+
+            if is_initiator
+              key_usage = Rex::Proto::Kerberos::Crypto::KeyUsage::GSS_INITIATOR_SEAL
+            else
+              key_usage = Rex::Proto::Kerberos::Crypto::KeyUsage::GSS_ACCEPTOR_SEAL
+              flags |= GSS_SENT_BY_ACCEPTOR
+            end
+
+            tok_id = 0x0504
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb:32]: We could add a new constant for this `TOK_ID_GSS_WRAP`

File: lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            filler = 0xFF
+            ec = calculate_ec(plaintext)
+            rrc = calculate_rrc
+
+            # RFC4121, Section 4.2.4
+            plaintext_header = [tok_id, flags, filler, 0, 0, sequence_number].pack('nCCnnQ>')
+
+            header = [tok_id, flags, filler, ec, rrc, sequence_number].pack('nCCnnQ>')
+            ec_filler = "x" * ec
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb:41]: @smashery Is there any context on this, is `x` the normal value here? 👀

File: lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+              key_usage = Rex::Proto::Kerberos::Crypto::KeyUsage::GSS_ACCEPTOR_SEAL
+              expected_flags |= GSS_SENT_BY_ACCEPTOR
+            else
+              key_usage = Rex::Proto::Kerberos::Crypto::KeyUsage::GSS_INITIATOR_SEAL
+            end
+            header = ciphertext[0,GSS_HEADER_LEN]
+            ciphertext = ciphertext[GSS_HEADER_LEN, ciphertext.length]
+
+            tok_id, flags, filler, ec, rrc, snd_seq = header.unpack('nCCnnQ>')
+            raise Rex::Proto::Kerberos::Model::Error::KerberosError, 'Invalid token ID' unless tok_id == 0x0504
```
Comment (adfoster-r7) [inline lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb:63]: We'd be able to swap this magic value `TOK_ID_GSS_WRAP`

File: lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb [hunk 5]
```diff
@@ -193,35 +221,108 @@ def authenticate(options = {})
...
+
+    checksum = Rex::Proto::Kerberos::Model::Checksum.new(
+      type: Rex::Proto::Gss::KRB_AP_REQ_CHKSUM_TYPE,
+      checksum: checksum_val)
+  end
+
   def build_service_ap_request(opts = {})
     authenticator = opts.fetch(:authenticator) do
       build_authenticator(opts.merge(
         subkey: nil,
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb:316]: @smashery Just confirming, were we expecting a subkey to be generated here?

File: lib/rex/proto/http/client.rb [hunk 6]
```diff
@@ -462,6 +478,58 @@ def digest_auth(opts={})
...
+    gss_data_b64 = Rex::Text.encode_base64(gss_data)
+    rescue => exception
+      print("Exception: #{exception}\n")
+      print("Exception: #{exception.backtrace}\n")
+    end
+
+    # Separate options for the auth requests
+    auth_opts = opts.clone
+    auth_opts['headers'] = opts['headers'].clone
+    auth_opts['headers']['Authorization'] = "Kerberos #{gss_data_b64}"
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/rex/proto/http/client.rb:495]: So in this case, it's in the `auth_kerberos` function, which is reached specifically in the case where `Authorization: Kerberos` is used. If SPNEGO were intended to be used, then the user of the HTTP Client object should specify `preferred_auth = Negotiate`, and it won't enter this function.

Ultimately, when using SPNEGO, we may need to be be more explicit about whether we want to use SPNEGO to negotiate NTLM or Kerberos - as of now, SPNEGO implies NTLM in the HTTP client.

File: lib/rex/proto/kerberos/model.rb (hunk 1 above)
Comment (smashery) [reply lib/rex/proto/kerberos/model.rb:11]: Yeah mostly on this one this was a note for later given all the merging I knew we'd have to do - wanted to reduce conflicts. Might just do a separate PR later.

Comment (adfoster-r7) [reply lib/rex/proto/kerberos/model.rb:11]: That's a good call, thanks for the clarification! :+1:

File: lib/rex/proto/http/client.rb [hunk 7]
```diff
@@ -462,6 +478,58 @@ def digest_auth(opts={})
...
+    begin
+      # Send the auth request
+      r = request_cgi(auth_opts)
+      resp = _send_recv(r, to)
+      unless resp.kind_of? Rex::Proto::Http::Response
+        return nil
+      end
+
+      # Get the challenge and craft the response
+      response = resp.headers['WWW-Authenticate'].scan(/Kerberos ([A-Z0-9\x2b\x2f=]+)/ni).flatten[0]
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/rex/proto/http/client.rb:512]: Comment as above - I think this is correct here, given it's within the `kerberos_auth` function.

File: lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb [hunk 8]
```diff
@@ -193,35 +221,108 @@ def authenticate(options = {})
 
     ## Service Authentication
 
+    checksum = nil
+    checksum = build_gss_ap_req_checksum_value(mutual_auth) if use_gss_checksum
+
+    sequence_number = rand(1 << 32)
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb:227]: So the sequence number in the GSS-wrapped messages for AES encryption is indeed 64 bits. However, the initial value is set in the Authenticator structure (https://datatracker.ietf.org/doc/html/rfc4120#section-5.5.1), and it is set to 32 bits. A reason for this is probably that older encryption types (i.e. RC4) actually only support 32-bit sequence numbers. That said, I'm yet to handle the wrapping of this value back around to zero once it reaches INT_MAX (which is within the realms of possibility for RC4; less likely for AES). I'll add that in as part of today's changes.

File: lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb (hunk 3 above)
Comment (smashery) [reply lib/rex/proto/kerberos/crypto/gss_new_encryption_type.rb:41]: I couldn't find an official definition of what this value should be. RFC4121 says "The
 values and size of the filler octets are chosen by implementations, such that there SHALL be no crypto-system residue present after the
 decryption." (https://datatracker.ietf.org/doc/html/rfc4121#section-4.2.4)

I used `'x'` based on the [(Linux implementation of the Kerberos client)](https://salsa.debian.org/debian/krb5/-/blob/0269810b1aec6c554fb746433f045d59fd34ab3a/src/lib/gssapi/krb5/k5sealv3.c#L160)

I'll make that clearer in a comment.

File: lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb (hunk 5 above)
Comment (smashery) [reply lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb:316]: So I don't believe it's required to supply a subkey; in its absence, the TGS-negotiated session key is used (although ultimately, the key provided in the AP-REP is used).

File: lib/msf/core/exploit/remote/kerberos/service_authenticator/http.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: binary -*-
+
+#
+# This class acts as standalone authenticator for Kerberos
+#
+class Msf::Exploit::Remote::Kerberos::ServiceAuthenticator::HTTP < Msf::Exploit::Remote::Kerberos::ServiceAuthenticator::Base
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/msf/core/exploit/remote/kerberos/service_authenticator/http.rb:6]: I think a HTTP-specific authenticator deserves to exist, as one could envisage other clients using Kerberos (whether directly or via SPNEGO) over HTTP that aren't related to WinRM. What do you think about a WinRM subclass that doesn't do much in addition to the HTTP one?

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msf/core/exploit/remote/kerberos/service_authenticator/http.rb:6]: Sounds good to me :+1: Either way we should probably default the mechanism to `Kerberos` for WinRM :+1:

**Review by smcintyre-r7 — APPROVED** [review]
Just tested this successfully on a few domain controllers. Everything seems to be working as intended.

* :green_circle: Server 2008 -- Properly negotiates type 23 encryption ARC4+MD5
* :green_circle: Server 2019 -- Properly negotiates type 18 encryption AES256+SHA1
* :green_circle: Server 2022 -- Properly negotiates type 18 encryption AES256+SHA1

<details>
<summary>Testing Output</summary>

## Server 2008

```
msf6 auxiliary(scanner/winrm/winrm_login) > resource test-kerberos-08.rc
[*] Processing /home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc for ERB directives.
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set RHOSTS 192.168.159.94
RHOSTS => 192.168.159.94
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set WinrmAuth kerberos
WinrmAuth => kerberos
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set WinrmRhostname WIN-RALP9EONK0T.msflab08.local
WinrmRhostname => WIN-RALP9EONK0T.msflab08.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set WinrmDomain msflab08.local
WinrmDomain => msflab08.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set DomainControllerRhost 192.168.159.94
DomainControllerRhost => 192.168.159.94
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set USERNAME smcintyre
USERNAME => smcintyre
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set PASSWORD Password1!
PASSWORD => Password1!
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set DOMAIN msflab08.local
DOMAIN => msflab08.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> set VERBOSE True
VERBOSE => true
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-08.rc)> run

[*] 192.168.159.94:88 - Received a valid TGT-Response
[+] 192.168.159.94:88 - Received a valid TGS-Response
[*] 192.168.159.94:88 - TGS MIT Credential Cache saved to /home/smcintyre/.msf4/loot/20220715105543_default_192.168.159.94_mit.kerberos.cca_558587.bin
[+] 192.168.159.94:88 - Received AP-REQ. Extracting session key...
[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.94:5985 - Login Successful: msflab08.local\smcintyre:Password1!
[*] Command shell session 11 opened (192.168.159.128:42375 -> 192.168.159.94:5985) at 2022-07-15 10:55:43 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) >
```

## Server 2019

```
msf6 auxiliary(scanner/winrm/winrm_login) > resource test-kerberos-19.rc
[*] Processing /home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc for ERB directives.
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set RHOSTS 192.168.159.10
RHOSTS => 192.168.159.10
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set WinrmAuth kerberos
WinrmAuth => kerberos
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set WinrmRhostname WIN-3MSP8K2LCGC.msflab.local
WinrmRhostname => WIN-3MSP8K2LCGC.msflab.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set WinrmDomain msflab.local
WinrmDomain => msflab.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set DomainControllerRhost 192.168.159.10
DomainControllerRhost => 192.168.159.10
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set USERNAME smcintyre
USERNAME => smcintyre
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set PASSWORD Password1!
PASSWORD => Password1!
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set DOMAIN msflab.local
DOMAIN => msflab.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> set VERBOSE True
VERBOSE => true
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-19.rc)> run

[*] 192.168.159.10:88 - Received a valid TGT-Response
[+] 192.168.159.10:88 - Received a valid TGS-Response
[*] 192.168.159.10:88 - TGS MIT Credential Cache saved to /home/smcintyre/.msf4/loot/20220715105407_default_192.168.159.10_mit.kerberos.cca_338406.bin
[+] 192.168.159.10:88 - Received AP-REQ. Extracting session key...
[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.10:5985 - Login Successful: msflab.local\smcintyre:Password1!
[*] Command shell session 10 opened (192.168.159.128:40005 -> 192.168.159.10:5985) at 2022-07-15 10:54:07 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) >
```

## Server 2022

```
msf6 auxiliary(scanner/winrm/winrm_login) > resource test-kerberos-22.rc
[*] Processing /home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc for ERB directives.
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set RHOSTS 192.168.159.46
RHOSTS => 192.168.159.46
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set WinrmAuth kerberos
WinrmAuth => kerberos
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set WinrmRhostname WIN-AEL58BHS5FJ.msflab22.local
WinrmRhostname => WIN-AEL58BHS5FJ.msflab22.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set WinrmDomain msflab22.local
WinrmDomain => msflab22.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set DomainControllerRhost 192.168.159.46
DomainControllerRhost => 192.168.159.46
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set USERNAME smcintyre
USERNAME => smcintyre
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set PASSWORD Password1!
PASSWORD => Password1!
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set DOMAIN msflab22.local
DOMAIN => msflab22.local
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> set VERBOSE True
VERBOSE => true
resource (/home/smcintyre/Repositories/metasploit-framework.pr/test-kerberos-22.rc)> run

[*] 192.168.159.46:88 - Received a valid TGT-Response
[+] 192.168.159.46:88 - Received a valid TGS-Response
[*] 192.168.159.46:88 - TGS MIT Credential Cache saved to /home/smcintyre/.msf4/loot/20220715104649_default_192.168.159.46_mit.kerberos.cca_240261.bin
[+] 192.168.159.46:88 - Received AP-REQ. Extracting session key...
[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.46:5985 - Login Successful: msflab22.local\smcintyre:Password1!
[*] Command shell session 9 opened (192.168.159.128:43837 -> 192.168.159.46:5985) at 2022-07-15 10:46:50 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) >
```

**Comment by adfoster-r7** [comment]
Hit some weirdness against my 2008 box when swapping `powershell` to `cmd`, but not a blocker as I've invalidated the warranty a bit there - even if it works against Windows Server 2016:

image

**Comment by adfoster-r7** [comment]
## Release Notes

Adds Kerberos Authentication support to WinRM modules

**Pull request merged**